### PR TITLE
Drau/layout

### PR DESCRIPTION
**Pull request by daverau-optimizely**
This PR adds a grid layout system to OUI that's fixed/fluid/responsive depending on what you need. It's versatile and will help us create more consistent layouts and spacing across our apps. It's based on Bootstrap's 12 column responsive grid implementation.

Update: I cherry picked the global Storybook search/replace capitalization into a [separate PR: #1180](https://github.com/optimizely/oui/pull/1180)

Screen Shot 2019-06-20 at 2 17 10 PM

# Done
1. [x] Establish baseline 12-col grid
1. [x] Remove unnecessary breakpoints
1. [x] Add Col padded prop
1. [x] Add Col border prop
1. [x] Add Row vertical column prop
1. [x] Add Viewports Storybook add-on
1. [x] Add border prop to Row
1. [x] Add padding prop to Row
1. [x] Refine colors for outline debug
1. [x] Consolidate basic examples

# Future component needs
- Add `height=100%` prop to Layout
- Add `Alignment` component
- Add `Overflow scroll` component

### CODE REVIEW

**Comment by davemn** [comment]
I didn't know we had a responsive grid planned ... this is exciting!

File: .storybook/config.js [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import * as storybook from '@storybook/react';
 import { withA11y } from '@storybook/addon-a11y';
-import { withInfo } from '@storybook/addon-info';
```
Comment (ayeshakmaz) [inline .storybook/config.js]: Think this was accidentally deleted, it's causing console errors, plus none of the stories have info anymore 😛

File: .storybook/config.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
...
+import { addParameters } from '@storybook/react';
```
Comment (ayeshakmaz) [inline .storybook/config.js]: Don't think this needs to be implicitly imported, `storybook.addParameters` should work ( like lines 40 and 43)

File: .storybook/config.js [hunk 3]
```diff
@@ -32,6 +33,7 @@ function loadStories() {
   require('./css.story.js');
   require('./borderradius.story.js');
   req.keys().forEach(req);
+  require('./AppFrame.story.js');
```
Comment (ayeshakmaz) [inline .storybook/config.js]: Is this needed if you're adding it to the Layout story?

File: src/components/Layout/Container.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const propTypes = {
+  /**
+   * Use a custom element for this component
+   */
+  as: PropTypes.elementType,
+  /**
+   * Allow the Container to fill availble horizontal space.
```
Comment (ayeshakmaz) [inline src/components/Layout/Container.js]: typo: available*

File: src/components/Layout/Container.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+   */
+  fluid: PropTypes.bool,
+  /**
+   * For dev only -- show hotpink border and background on grid elements.
```
Comment (ayeshakmaz) [inline src/components/Layout/Container.js]: maybe "For testing/debugging only"?

File: src/components/Layout/Layout.story.js [hunk 6]
```diff
@@ -0,0 +1,566 @@
+import React from 'react';
+
+import { storiesOf } from '@storybook/react';
+import { withKnobs, boolean, number, select } from '@storybook/addon-knobs';
+import { withInfo } from '@storybook/addon-info';
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: Shouldn't need withInfo here, and instead rely on the global config to apply it (can also delete all the withInfo() additions for each story in this file)

File: src/components/Layout/Layout.story.js [hunk 7]
```diff
@@ -0,0 +1,566 @@
...
+  ipad: {
+    name: 'Tablet',
+    styles: {
+      height: '1024px',
+      width: '768px',
+    },
+    type: 'tablet',
+  },
+};
+addParameters({ viewport: { viewports: viewports } });
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: I don't see these coming through, but I might be missing something

File: src/components/Layout/Col.js [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  /**
+   * The number of columns to span on large devices (≥992px)
+   *
+   * @type {("fillSpace"|"fitContent"|number|
+   *  { span: "fillSpace"|"fitContent"|number, offset: number, order: number }
+   * )}
+   */
+  large: column,
+
+  paddedContent: PropTypes.oneOf([
```
Comment (ayeshakmaz) [inline src/components/Layout/Col.js]: should 'none' be an option here? I'm seeing a console error

File: src/components/Layout/README.md [hunk 9]
```diff
@@ -0,0 +1 @@
+# Flexbox Layout Component
```
Comment (ayeshakmaz) [inline src/components/Layout/README.md]: Are these files needed anymore?

File: src/components/Layout/Row.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+const propTypes = {
+  /**
+   * @default 'row'
+   */
+  as: PropTypes.elementType,
+  /**
+   * Add visible border to row
+  */
+  border: PropTypes.oneOf([
```
Comment (ayeshakmaz) [inline src/components/Layout/Row.js]: Should `none` be an option here? Seeing a console error for propTypes

File: src/components/Layout/Row.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    'sides',
+    'ends',
+    'all',
+  ]),
+  /**
+   * Display layout as vertical column.
+  */
+  displayVertical: PropTypes.bool,
+  /**
+   * For dev only -- show hotpink border and background on grid elements.
```
Comment (ayeshakmaz) [inline src/components/Layout/Row.js]: Same comment here about the wording

File: src/components/Layout/ThemeProvider.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
+import PropTypes from 'prop-types';
```
Comment (ayeshakmaz) [inline src/components/Layout/ThemeProvider.js]: Are we using this ThemeProvider?

File: src/components/Layout/Layout.story.js [hunk 13]
```diff
@@ -0,0 +1,566 @@
...
+        <Row>
+          <Col small={ 'fitContent' }>fitContent</Col>
+          <Col small={ 'fitContent' }>fitContent</Col>
+          <Col small={ 'fitContent' }>fitContent</Col>
+        </Row>
+      </Container>
+    </div>
+  );
+}))
+  .add('Minimal', withInfo()(() => {
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: Might be good to be a little more verbose on these section names, so maybe:
`Container options` or `Container differences`

File: src/components/Layout/Layout.story.js [hunk 14]
```diff
@@ -0,0 +1,566 @@
...
+      </Container>
+      <h1 className="push-triple--top">No Container</h1>
+      <Row
+        outlineDebug={ boolean('outlineDebug', true) }>
+        <Col paddedContent={ select('paddedContent', paddingOptions, 'none') }>Hello short example</Col>
+        <Col paddedContent={ select('paddedContent', paddingOptions, 'none') }>Hello short example</Col>
+      </Row>
+    </div>);
+  }))
+  .add('Padding', withInfo()(() => {
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: Maybe: `Column padding`

File: src/components/Layout/Layout.story.js [hunk 15]
```diff
@@ -0,0 +1,566 @@
...
+            large={ 'fillSpace' }
+            paddedContent={ select('paddedContent', paddingOptions, 'around') }
+            border={ select('Col border', borderOptions, 'all') }>
+            Change the padding on individual Columns via props. Try the knob to adjust padding in these cells together.
+          </Col>
+        </Row>
+      </Container>
+    );
+  }))
+  .add('Borders', withInfo()(() => {
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: `Border options`?

File: src/components/Layout/Layout.story.js [hunk 16]
```diff
@@ -0,0 +1,566 @@
...
+          <Col
+            paddedContent={ 'around' }
+            border={ 'left' }>
+            Col with border right
+          </Col>
+        </Row>
+      </Container>
+    );
+  }))
+  .add('Pull', withInfo()(() => {
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: maybe `Row padding options with Pull`?

File: src/components/Layout/Layout.story.js [hunk 17]
```diff
@@ -0,0 +1,566 @@
...
+          <Col>pull removes left padding here.</Col>
+          <Col>pull does nothing here.</Col>
+          <Col>pull does nothing here.</Col>
+          <Col>pull does nothing here.</Col>
+          <Col>pull removes right padding here.</Col>
+        </Row>
+      </Container>
+    );
+  }))
+  .add('Vertical', withInfo()(() => {
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: `Vertical layout example`?

File: src/components/Layout/Layout.story.js [hunk 18]
```diff
@@ -0,0 +1,566 @@
...
+        <Row
+          removeGutters={ boolean('Row removeGutters', false) }
+          displayVertical={ boolean('displayVertical', true) }>
+          <Col>1 of 2</Col>
+          <Col>2 of 2</Col>
+        </Row>
+      </Container>
+    );
+  }))
+  .add('Order', withInfo()(() => {
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: `Changing column order`?

File: src/components/Layout/Layout.story.js [hunk 19]
```diff
@@ -0,0 +1,566 @@
...
+        fluid={ boolean('fluid', false) }>
+        <Row removeGutters={ boolean('Row removeGutters', false) }>
+          <Col small={ true }>First, but unordered</Col>
+          <Col small={{ order: 12 }}>Second, but last</Col>
+          <Col small={{ order: 1 }}>Third, but second</Col>
+        </Row>
+      </Container>
+    );
+  }))
+  .add('Widths', withInfo()(() => {
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: Little confused what this example is showing and how the `1 of 3` `2 of 3` and `3 of 3` column contents relate

File: src/components/Layout/Layout.story.js [hunk 20]
```diff
@@ -0,0 +1,566 @@
...
+        fluid={ boolean('fluid', false) }>
+        <Row>
+          <Col>1 of 3</Col>
+          <Col small={ number('number', 6) }>2 of 3 (variable)</Col>
+          <Col>3 of 3</Col>
+        </Row>
+      </Container>
+    );
+  }))
+  .add('Cards', withInfo()(() => {
```
Comment (ayeshakmaz) [inline src/components/Layout/Layout.story.js]: `Using the Card component`?

File: src/components/Layout/Container.js [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+    'none',
+    'around',
+    'sides',
+    'ends',
+    'remove',
+  ]),
+  /**
+   * Remove first and last child side padding.
+   */
+  pull: PropTypes.bool,
```
Comment (ayeshakmaz) [inline src/components/Layout/Container.js]: I'm not sure this prop name is clear enough, maybe `pullRowPadding` or something similar?

File: src/components/Layout/Col.js [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+    'left',
+    'right',
+    'sides',
+    'ends',
+    'all',
+  ]),
+
+  children: PropTypes.node,
+
+  className: PropTypes.string,
```
Comment (ayeshakmaz) [inline src/components/Layout/Col.js]: Do we want people to be able to add any className to `Col`?

File: src/components/Layout/Col.js [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+      'sides',
+      'ends',
+      'remove',
+    ]),
+    size: colSize,
+  }),
+]);
+
+const propTypes = {
+  as: PropTypes.elementType,
```
Comment (ayeshakmaz) [inline src/components/Layout/Col.js]: Can you add propType descriptions for all the props here? There's a few missing

File: src/components/Layout/Col.js (hunk 22 above)
Comment (daverau-optimizely) [reply src/components/Layout/Col.js]: Oops, this was removed but I think it got eaten when I was doing merge conflict stuff.

File: src/components/Layout/Col.js (hunk 8 above)
Comment (daverau-optimizely) [reply src/components/Layout/Col.js]: Hmm, interesting thought, I wasn't picturing that here. Would you imagine none would hide the column at this size?

File: src/components/Layout/README.md (hunk 9 above)
Comment (daverau-optimizely) [reply src/components/Layout/README.md]: Nope!

File: src/components/Layout/Row.js (hunk 10 above)
Comment (daverau-optimizely) [reply src/components/Layout/Row.js]: Good catch.

File: src/components/Layout/ThemeProvider.js (hunk 12 above)
Comment (daverau-optimizely) [reply src/components/Layout/ThemeProvider.js]: It's baked in, so there's work involved to remove it and I was trying to push it out there first and see if this is something we'd consider later. Potential darkmode or something else, idk...

File: src/components/Layout/Layout.story.js (hunk 20 above)
Comment (daverau-optimizely) [reply src/components/Layout/Layout.story.js]: Good suggestion

File: src/components/Layout/Layout.story.js (hunk 19 above)
Comment (daverau-optimizely) [reply src/components/Layout/Layout.story.js]: This shows adjustable widths via the `small=` knob. I updated the example story text to be more clear and focused:

Screen Shot 2019-06-26 at 5 36 37 PM

**Comment by daverau-optimizely** [comment]
Thanks @ayeshakmaz — great review! 💯 I appreciate that you ran this locally and tested against other props overlooked (like `none` for example).

File: .storybook/config.js (hunk 3 above)
Comment (daverau-optimizely) [reply .storybook/config.js]: Yes, I tried to add it via the Layout story but that includes extra wrapper output I don't want for App Frame.

Also I tried to move this into the Layout folder itself and ran into issues where the app frame story was overwriting the other Layout story... so I gave up and just kept this since it's working 🙃

File: src/components/Layout/Col.js (hunk 23 above)
Comment (daverau-optimizely) [reply src/components/Layout/Col.js]: Done!
